File: src/services/users.service.ts
import bcrypt from 'bcryptjs';
import { queryOne, execute } from '../config/db';
import { fileService } from './file.service';

function formatUser(row: any, baseUrl = '') {
  return {
    id:          row.id,
    name:        row.name,
    email:       row.email,
    phone:       row.phone,
    // Resolve avatar_path → serve URL; fall back to avatar_url for external URLs
    avatar_url:  row.avatar_path
                   ? fileService.serveUrl(row.avatar_path, baseUrl)
                   : (row.avatar_url ?? null),
    created_at:  row.created_at instanceof Date ? row.created_at.toISOString() : row.created_at,
  };
}

export const usersService = {
  // ── Public profile ────────────────────────────────────────────────────────
  async getPublicProfile(id: string, baseUrl = '') {
    const user = await queryOne<any>('SELECT * FROM users WHERE id = ?', [id]);
    if (!user) throw Object.assign(new Error('User not found.'), { status: 404, code: 'not_found' });

    const countRow = await queryOne<{ cnt: number }>(
      'SELECT COUNT(*) AS cnt FROM phone_ads WHERE user_id = ? AND is_active = 1',
      [id]
    );

    return {
      id:         user.id,
      name:       user.name,
      avatar_url: user.avatar_path
                    ? fileService.serveUrl(user.avatar_path, baseUrl)
                    : (user.avatar_url ?? null),
      created_at: user.created_at instanceof Date ? user.created_at.toISOString() : user.created_at,
      ad_count:   countRow?.cnt || 0,
    };
  },

  // ── Update profile (name / phone) ─────────────────────────────────────────
  async updateProfile(userId: string, data: { name?: string; phone?: string }, baseUrl = '') {
    if (data.phone) {
      const conflict = await queryOne<any>(
        'SELECT id FROM users WHERE phone = ? AND id != ?',
        [data.phone, userId]
      );
      if (conflict) throw Object.assign(new Error('This phone number is already in use.'), { status: 409, code: 'phone_taken' });
    }

    const fields: string[] = [];
    const values: any[]    = [];

    if (data.name  !== undefined) { fields.push('name = ?');  values.push(data.name.trim()); }
    if (data.phone !== undefined) { fields.push('phone = ?'); values.push(data.phone); }

    if (fields.length > 0) {
      fields.push('updated_at = ?');
      values.push(new Date());
      values.push(userId);
      await execute(`UPDATE users SET ${fields.join(', ')} WHERE id = ?`, values);

      // Sync seller_name on active ads if name changed
      if (data.name) {
        await execute(
          'UPDATE phone_ads SET seller_name = ?, updated_at = ? WHERE user_id = ? AND is_active = 1',
          [data.name.trim(), new Date(), userId]
        );
      }
    }

    const user = await queryOne<any>('SELECT * FROM users WHERE id = ?', [userId]);
    return formatUser(user, baseUrl);
  },

  // ── Update avatar (upload file → store blob → save path on user) ──────────
  async updateAvatar(
    userId: string,
    file: Express.Multer.File,
    baseUrl = ''
  ) {
    // Delete old avatar blob if one exists
    const existing = await queryOne<any>('SELECT avatar_path FROM users WHERE id = ?', [userId]);
    if (existing?.avatar_path) {
      await fileService.deleteByPath(existing.avatar_path).catch(() => { /* ignore if missing */ });
    }

    // Save new blob to files table
    const saved = await fileService.save({
      originalName: file.originalname,
      mimeType:     file.mimetype,
      size:         file.size,
      buffer:       file.buffer,
      module:       'users',
      uploadedBy:   userId,
    });

    // Store path on user row (avatar_path = e.g. "users/1714123456_avatar.jpg")
    await execute('UPDATE users SET avatar_path = ?, updated_at = ? WHERE id = ?', [saved.path, new Date(), userId]);

    const user = await queryOne<any>('SELECT * FROM users WHERE id = ?', [userId]);
    return formatUser(user, baseUrl);
  },

  // ── Change password ───────────────────────────────────────────────────────
  async changePassword(userId: string, currentPassword: string, newPassword: string) {
    const user = await queryOne<any>('SELECT * FROM users WHERE id = ?', [userId]);
    if (!user) throw Object.assign(new Error('User not found.'), { status: 404, code: 'not_found' });

    const valid = await bcrypt.compare(currentPassword, user.password_hash);
    if (!valid) throw Object.assign(new Error('Current password is incorrect.'), { status: 400, code: 'invalid_credentials' });

    const hashed = await bcrypt.hash(newPassword, 12);
    await execute('UPDATE users SET password_hash = ? WHERE id = ?', [hashed, userId]);
  },
};

File: EasyPhones.Application/Services/UsersService.cs
using EasyPhones.Application.Dtos.AuthDtos;
using EasyPhones.Application.Dtos.UsersDtos;
using EasyPhones.Application.Exceptions;
using EasyPhones.Domain.Repositories;
using BCrypt.Net;

namespace EasyPhones.Application.Services;

public class UsersService
{
    private readonly IUserRepository _users;
    private readonly IAdRepository _ads;
    private readonly IFileService _files;

    public UsersService(IUserRepository users, IAdRepository ads, IFileService files)
    {
        _users = users;
        _ads = ads;
        _files = files;
    }

    // ── Public profile ────────────────────────────────────────────────────────
    public async Task<PublicProfileDto> GetPublicProfileAsync(string id, string baseUrl, CancellationToken ct)
    {
        var user = await _users.GetByIdAsync(id, ct)
            ?? throw ApiException.NotFound("User not found.");

        var userAds = await _ads.GetByUserAsync(id, ct);
        var activeCount = userAds.Count(a => a.IsActive);

        var avatarUrl = user.AvatarPath is not null
            ? _files.ServeUrl(user.AvatarPath, baseUrl)
            : user.AvatarUrl;

        return new PublicProfileDto
        {
            Id = user.Id,
            Name = user.Name,
            AvatarUrl = avatarUrl,
            CreatedAt = user.CreatedAt.ToString("O"),
            AdCount = activeCount
        };
    }

    // ── Update profile ────────────────────────────────────────────────────────
    public async Task<UserProfileDto> UpdateProfileAsync(string userId, UpdateProfileDto dto, string baseUrl, CancellationToken ct)
    {
        if (dto.Phone is not null)
        {
            var conflict = await _users.GetByPhoneAsync(dto.Phone, ct);
            if (conflict is not null && conflict.Id != userId)
                throw ApiException.Conflict("This phone number is already in use.", "phone_taken");
        }

        var user = await _users.GetByIdAsync(userId, ct)
            ?? throw ApiException.NotFound("User not found.");

        if (dto.Name is not null)  user.Name = dto.Name.Trim();
        if (dto.Phone is not null) user.Phone = dto.Phone;
        user.UpdatedAt = DateTime.UtcNow;

        await _users.UpdateAsync(user, ct);

        var avatarUrl = user.AvatarPath is not null
            ? _files.ServeUrl(user.AvatarPath, baseUrl)
            : user.AvatarUrl;

        return new UserProfileDto
        {
            Id = user.Id,
            Name = user.Name,
            Email = user.Email,
            Phone = user.Phone,
            AvatarUrl = avatarUrl,
            CreatedAt = user.CreatedAt.ToString("O")
        };
    }

    // ── Update avatar ─────────────────────────────────────────────────────────
    public async Task<UserProfileDto> UpdateAvatarAsync(
        string userId, string fileName, string contentType, long size, byte[] bytes,
        string baseUrl, CancellationToken ct)
    {
        var user = await _users.GetByIdAsync(userId, ct)
            ?? throw ApiException.NotFound("User not found.");

        if (user.AvatarPath is not null)
            await _files.DeleteByPathAsync(user.AvatarPath, ct).ContinueWith(_ => { });

        var saved = await _files.SaveAsync(new FileUploadInput(fileName, contentType, size, bytes, "users", userId), ct);
        user.AvatarPath = saved.Path;
        user.UpdatedAt = DateTime.UtcNow;
        await _users.UpdateAsync(user, ct);

        return new UserProfileDto
        {
            Id = user.Id,
            Name = user.Name,
            Email = user.Email,
            Phone = user.Phone,
            AvatarUrl = _files.ServeUrl(saved.Path, baseUrl),
            CreatedAt = user.CreatedAt.ToString("O")
        };
    }

    // ── Change password ───────────────────────────────────────────────────────
    public async Task ChangePasswordAsync(string userId, ChangePasswordDto dto, CancellationToken ct)
    {
        if (dto.NewPassword != dto.ConfirmPassword)
            throw ApiException.Validation("Passwords do not match.");

        var user = await _users.GetByIdAsync(userId, ct)
            ?? throw ApiException.NotFound("User not found.");

        if (!BCrypt.Net.BCrypt.Verify(dto.CurrentPassword, user.PasswordHash))
            throw ApiException.Validation("Current password is incorrect.", "invalid_credentials");

        user.PasswordHash = BCrypt.Net.BCrypt.HashPassword(dto.NewPassword, 12);
        user.UpdatedAt = DateTime.UtcNow;
        await _users.UpdateAsync(user, ct);
    }
}
